<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Things you will learn in this course</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />
<!-- Global site tag (gtag.js) - Google Analytics -->
<script async src="https://www.googletagmanager.com/gtag/js?id=UA-39623718-2"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-39623718-2');
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>


<link rel="stylesheet" href="libs/style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">
    <span class="fas fa-home"></span>
     
    
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Introduction
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Week01.html">Course Introduction</a>
    </li>
    <li>
      <a href="The_R_environment.html">The R and RStudio environments</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Data Manipulation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Week 02</li>
    <li>
      <a href="Week02a.html">Data types</a>
    </li>
    <li>
      <a href="Week02b.html">Reading and writing data files</a>
    </li>
    <li>
      <a href="Week02c.html">Working with date objects</a>
    </li>
    <li>
      <a href="Week02d.html">Exploring and cleaning dataframes using base functions</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 03</li>
    <li>
      <a href="Week03a.html">Manipulating dataframes with dplyr</a>
    </li>
    <li>
      <a href="Week03d.html">Working with string objects</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 04</li>
    <li>
      <a href="Week03ab_groupby.html">Grouping and summarizing tables</a>
    </li>
    <li>
      <a href="Week03b.html">Tidying/reshaping tables using tidyr</a>
    </li>
    <li>
      <a href="Week03c.html">Joining data tables</a>
    </li>
    <li>
      <a href="Week03ab_examples.html">Example of data manipulation workflows</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Plots
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Week 05</li>
    <li>
      <a href="Week04a.html">The base plotting environment</a>
    </li>
    <li>
      <a href="Week04b.html">The lattice plotting environment</a>
    </li>
    <li>
      <a href="Week04c.html">The ggplot plotting environment</a>
    </li>
    <li>
      <a href="Week04d.html">Manipulating colors in R</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Univariate
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Week 06</li>
    <li>
      <a href="Week05a.html">Visualizing univariate data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 07</li>
    <li>
      <a href="Week05b.html">Comparing univariate data distributions</a>
    </li>
    <li>
      <a href="Week06a.html">Theoretical Q-Q plot</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 08</li>
    <li>
      <a href="Week07a.html">Fits and residuals</a>
    </li>
    <li>
      <a href="Week07b.html">Spread-location plot</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 09</li>
    <li>
      <a href="Week08a.html">Re-expressing data</a>
    </li>
    <li>
      <a href="Week08b.html">Letter value summaries</a>
    </li>
    <li>
      <a href="Week08c.html">The Two R’s of EDA</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Bivariate
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Week 10</li>
    <li>
      <a href="Week09a.html">Bivariate analysis</a>
    </li>
    <li>
      <a href="Week09b.html">Resistant lines</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Week 11</li>
    <li>
      <a href="Week10a.html">The third R of EDA: Residuals</a>
    </li>
    <li>
      <a href="Week10b.html">Detecting discontinuities in the data</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Two-way tables
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Week 12</li>
    <li>
      <a href="Week11a.html">Median polish/Mean polish</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Misc
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Misc01.html">R markdown documents</a>
    </li>
    <li class="dropdown-header">Week 13</li>
    <li>
      <a href="Week12a.html">Creating maps in R</a>
    </li>
    <li class="dropdown-header">Connecting to relational databases</li>
  </ul>
</li>
<li>
  <a href="Data.html">Datasets</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<link href="https://fonts.googleapis.com/css?family=Open+Sans%7CSource+Code+Pro" rel="stylesheet">

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Things you will learn in this course</h1>

</div>


<div id="exploratory-data-analysis-eda" class="section level1">
<h1>Exploratory Data Analysis (EDA)</h1>
<div id="what-is-eda" class="section level2">
<h2>What is EDA?</h2>
<p>Traditional approaches to data analysis tend to be linear and unidirectional. It often starts with the acquisition or collection of a data set and ends with the computation of some inferential or confirmatory procedure.</p>
<p><img src="img/week1_fig1.png" /></p>
<p>Unfortunately, such practice can lead to faulty conclusions. The following datasets generate identical regression analysis results shown in the previous figure yet, they are all completely different!</p>
<p><img src="Week01_files/figure-html/anscombe-1.png" width="624" /></p>
<p>The four plots represent Francis Anscombe’s famous quartet which he used to demonstrate the importance of visualizing the data before proceeding with traditional statistical analysis. Of the four plots, only the first is a sensible candidate for the regression analysis; the second dataset highlights a nonlinear relationship between X and Y; the third and fourth plots demonstrate the disproportionate influence of a single outlier on the regression procedure.</p>
<p>The aforementioned example demonstrates that a sound data analysis workflow must involve data visualization and exploration techniques. Exploratory data analysis seeks to extract salient features about the data (that may have otherwise gone unnoticed) and to help formulate hypotheses. Only then should appropriate statistical tests be applied to the data to confirm a hypothesis.</p>
<p>However, not all EDA workflows result in a statistical test: We may not be seeking a hypothesis or, if a hypothesis is sought we may not have the statistical tools necessary to test the hypothesis. It’s important to realize that most statistical procedures make restrictive assumptions about the data and the type of hypothesis being tested; data sets seldom meet those stringent requirements.</p>
<div style="width:500px;height:160px;margin-left:70px;margin-bottom:10px;;font-family:Garamond, Georgia, serif;font-size:1.5em;font-style:italic">
<p><img style="float:left;margin-right:10px;" src="img/Tukey.png"> “Exploratory data analysis is an attitude, a flexibility, and a reliance on display, NOT a bundle of techniques.”</p>
<p>–John Tukey</p>
</div>
<p>John Tukey is credited with having coined the term exploratory data analysis and with having written the first comprehensive book (Tukey, 1977<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>) on that subject in 1977. The book is still very much relevant today and several of the techniques highlighted in the book will be covered in this course.</p>
</div>
<div id="an-eda-example-co2-analysis" class="section level2">
<h2>An EDA example: CO2 analysis</h2>
<p>Let’s start off by plotting the atmospheric CO2 concentrations (in ppm) pulled from <a href="http://www.esrl.noaa.gov/gmd/ccgg/trends/">NOAA’s website</a></p>
<p><img src="Week01_files/figure-html/unnamed-chunk-2-1.png" width="576" /></p>
<p>We note two patterns of interest: an overall upward trend, and a cyclical trend. Our first EDA task is to model the overall trend. We can attempt to fit a straight line to the data using a standard regression analysis procedure. The fitted line is displayed in red in the following plot.</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-3-1.png" width="576" /></p>
<p>Next, we subtract the modeled line from the CO2 data and plot the result–this difference is called the <strong>residual</strong> and can be thought of as representing what the linear model does <em>not</em> explain about the data.</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-4-1.png" width="576" /></p>
<p>An overall trend is still present, despite having attempted to control for it. This implies that our simple line model does not do a good job in <em>smoothing</em> out the overall trend. It appears that the overall trend is slightly convex and has a small peak around the 1990’s; we should try to fit the trend using a 3rd order polynomial of the form: <span class="math display">\[
CO2_{trend} = a + b(Year) + c(Year^2) + d(Year^3)
\]</span></p>
<p>The fitted line looks like this:</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-5-1.png" width="576" /></p>
<p>Now, let’s look at the residuals:</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-6-1.png" width="576" /></p>
<p>This is an improvement over the simple line (which was a 1st order polynomial fit). So what we have learned so far is that the overall trend is not perfectly linear but instead follows a parabolic like trajectory with a small hump halfway across the time span. However, we can still make out a “W” shaped trend in the residual which can hamper our analysis of the oscillatory patterns in the data. We could play with different order polynomials to smooth the trend even further, but at this point, we may opt for a non-parametric fit. When the goal is to peel off one pattern to explore any underlying pattern we should not limit ourselves to parametric fits (which impose a mathematical model on our data) and instead explore non-parametric smoothing techniques that do not impose any structure on the data. An example of a non-parametric smoothing technique is the LOESS curve which is shown in the following figure.</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-7-1.png" width="576" /></p>
<p>At first glance, this may not look any different from our 3rd order polynomial. But the resulting residual suggests that the LOESS smooth did a good job in removing any overall trend in our batch of CO2 values.</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-8-1.png" width="576" /></p>
<p>Let’s now focus on the residuals. Note the y axis values: they are three orders of magnitude less than the overall CO2 values. This indicates that the oscillating nature about the overall trend is relatively small (the CO2 values have a range of [313.26, 411.24] whereas the residuals have a range of [-4.58, 4.37].</p>
<p>Now, we may be tempted to fit a <em>smooth</em> (as Tukey would say) to the residuals but that may not prove to be fruitful. Instead let’s see if the oscillation follows a 12 month cycle. We’ll group all the (residual) values by month of the year. In other words, we will remove the year <em>tag</em> associated with each value and explore those values as a function of month alone. Each month’s batch of values is distilled into boxplots (one for each month of the year).</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-9-1.png" width="384" /></p>
<p>The box plot is a nifty visualization tool used to distill a batch of numbers into a measure of centrality and spread. The box in the graphic encompasses 50% of the data; the horizontal line in the box represents the median (a measure of centrality); the vertical lines above and below the box, the “whiskers”, encompass most of the remaining data; the black points extending beyond the “whiskers” are deemed outliers (i.e. values that are not typical of the batch).</p>
<p>It’s clear from the plot that the oscillation we observed in the CO2 plot follows a yearly cycle: a peak in the spring and a dip in the fall. This cycle is explained in part by the increased land mass in the northern hemisphere relative to the southern hemisphere. Because plants (and by extension photosynthesis) goes dormant during the winter months in the northern hemisphere, atmospheric CO2 is no longer being photosynthesized; this despite the southern hemisphere’s photosynthesis peak during the October-March period (a result of the southern hemisphere’s smaller land mass). Other factors such as increased CO2 emissions during the winter months may also contribute to the oscillatory nature of atmospheric CO2 concentrations.</p>
<p>Thus far, we have uncovered two patterns of interest: an overall trend and a cyclical component. Note that to effectively explore the cyclical pattern we had to <strong>de-trend</strong> (or <strong>smooth</strong>) the data. Next, we should <em>smooth</em> the seasonal component of the data to see if another pattern emerges. We may, for example, smooth the data by subtracting the monthly median from each residual leaving us with the next batch of residual values to explore:</p>
<p><img src="Week01_files/figure-html/unnamed-chunk-10-1.png" width="384" /></p>
<p>Note that now, all the boxplots are lined up along their median values. We are now exploring the data after having accounted for both overall trend and seasonality. What can be gleaned from this dataset? We may want to explore the skewed nature of the residuals in February, or the narrower range of CO2 values for the fall months, for example.</p>
<p>It’s important to realize that this is one of many ways in which we could have proceeded with the analysis. For example, we could have started off the analysis by removing the seasonal component from the data, then analyze the long term trend. This is the approach taken by William Cleveland <a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<p>Data analysis is an interplay between fitting the data and plotting the residuals. The goal is to extract all meaningful patterns from the data and to be left with residuals that are randomly distributed around a value of 0. Some may caution that this is tantamount to <em>data snooping</em> or <em>forcing the data</em> to reveal what we want it to reveal. While there might be some truth to this, think of the alternative: had we not graphically inspected the CO2 data, how could we have uncovered all of the interesting features of the data? And without knowing how the data behaves, how would we know what statistical procedures would be adequate for the analysis?</p>
</div>
<div id="the-role-of-graphics-in-eda" class="section level2">
<h2>The role of graphics in EDA</h2>
<p>The preceding example highlights the importance of graphing data. A core component of this course is learning how to construct effective data visualization tools for the purpose of revealing patterns in the data. The graphical tools must allow the data to <em>express themselves</em> without imposing a <em>story</em>.</p>
<div style="width:500px;height:160px;margin-left:70px;margin-bottom:10px;font-family:Garamond, Georgia, serif;font-size:1.5em;font-style:italic">
<p><img style="float:left;margin-right:10px;" src="img/Cleveland.jpg"> “Visualization is critical to data analysis. It provides a front line of attack, revealing intricate structure in data that cannot be absorbed in any other way.”</p>
<p>–William S. Cleveland</p>
</div>
<p>William Cleveland has written extensively about data visualization and has focused on principles founded in the field of cognitive neuroscience to improve data graphic designs. His book, <em>Visualizing Data</em>, is a leading authority on statistical graphics and, despite its age, is as relevant today as it was two decades ago. It focuses on graphical techniques (some newer than others) designed to explore the data. This may differ from graphics generated for public dissemination which benefits from another form of data visualization called <em>information visualization</em> (or infovis for short). Infovis will not be covered in this course (though there is some overlap between the two techniques). For a good discussion on the differences between statistical graphics and infovis see the 2011 Statistical Computing and Graphics Newsletter <a href="http://stat-computing.org/newsletter/issues/scgn-22-1.pdf">Statistical Graphics and InfoVis: separated Twins at Birth?</a><a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a></p>
<p>Cleveland has also contributed a very important tool to EDA: the <em>LOESS</em> curve (which was demonstrated in the CO2 data example). The LOESS curve will be used extensively in this course.</p>
<p>The LOESS is one of many fitting options used in smoothing (or detrending) the data. Others include parametric models such as the family of linear polynomials (we explored two of them in the CO2 example) and Tukey’s suite of smoothers notably the <em>running median</em> and the <em>3RS3R</em>.</p>
</div>
</div>
<div id="we-need-a-good-data-analysis-environment" class="section level1">
<h1>We need a good data analysis environment</h1>
<p>Effective EDA requires a flexible data analysis environment that does not constrain one to a limited set of data manipulation procedures or visualization tools. After all, would any good writer limit herself to a set of a hundred pre-built sentences? Of course not–we would be reading the same novels over and over again! So why would we limit ourselves to a limited set of pre-packaged data analysis procedures? EDA requires an arsenal of data analysis building blocks much like a good writer needs an arsenal of words. Such an environment must provide us with flexible data manipulation capabilities, a flexible data visualization environment and access to a wide range of statistical procedures. A scripting environment, like R, offers such an environment.</p>
<p>The data analysis environment should also be freely available, and its code open to the public. Free access to the software allows anyone with the right set of skills to share in the data analysis, regardless of any budgetary constraints. The open source nature of the software ensures that any aspect of the code used for a particular task can be examined when additional insight into the implementation of an analytical/numerical method if needed. Deciphering code may not be a skill available to all researchers, however, if the need to understand how a procedure is implemented is important enough, an individual with the appropriate programming skills can be easy to come by, even if it’s for a small fee. Open source software also ensures that the underlying code used to create the executable application can be ported to different platforms or different operating systems (even though this too may require some effort and modest programming skills).</p>
<div id="the-workhorse-r" class="section level2">
<h2>The workhorse: R</h2>
<p><a href="https://en.wikipedia.org/wiki/R_%28programming_language%29">R</a> is an open source data analysis and visualization programming environment whose roots go back to the <a href="https://en.wikipedia.org/wiki/S_%28programming_language%29">S programming language</a> developed at Bell Laboratories in the 1970’s by <a href="https://en.wikipedia.org/wiki/John_Chambers_%28statistician%29">John Chambers</a>. It will be used almost exclusively in this course.</p>
</div>
<div id="the-friendly-interface-rstudio" class="section level2">
<h2>The friendly interface: RStudio</h2>
<p>RStudio is an integrated development environment (IDE) to R. An IDE provides a user with an interface to a programming environment (like R) by including features such as a source code editor (with colored syntax). RStudio is not needed to use R (which has its own IDE environment–albeit not as nice as RStudio’s), but makes using R far easier. RStudio is an open source software, but unlike R, it’s maintained by a private entity which also distributes a commercial version of RStudio for businesses or individuals needing customer support.</p>
</div>
<div id="data-manipulation" class="section level2">
<h2>Data manipulation</h2>
<p>The importance of data visualization was already discussed in an earlier section. But before one can begin plotting data, one must have a data table in a form ready to be plotted. In cases where the data table consists of just two variables (columns), little data manipulation may be needed, but in cases where data tables consist of tens or scores of variables, data manipulation, subsetting and/or reshaping may be required. Tackling such a task can be challenging in a point and click spreadsheet environment and can introduce clerical error. R offers an array of data table manipulation tools and packages such as <code>tidyr</code> and <code>dplyr</code>. Furthermore, R’s scripting environment enables one to <em>read</em> through each step of a manipulation procedure in a clear and unambiguous way. Imagine the difficulty in properly documenting all the point-and-click steps followed in a spreadsheet environment.</p>
For example, a data table of grain production for North America may consist of six variables and 1501 rows. The following table shows just the first 7 lines of the 1501 rows.
<table>
<thead>
<tr>
<th style="text-align:left;">
Country
</th>
<th style="text-align:left;">
Crop
</th>
<th style="text-align:left;">
Information
</th>
<th style="text-align:right;">
Year
</th>
<th style="text-align:right;">
Value
</th>
<th style="text-align:left;">
Source
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Barley
</td>
<td style="text-align:left;">
Area harvested (Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
2060000.00
</td>
<td style="text-align:left;">
Official data
</td>
</tr>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Barley
</td>
<td style="text-align:left;">
Yield (Hg/Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
38894.66
</td>
<td style="text-align:left;">
Calculated data
</td>
</tr>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Buckwheat
</td>
<td style="text-align:left;">
Area harvested (Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
0.00
</td>
<td style="text-align:left;">
FAO estimate
</td>
</tr>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Canary seed
</td>
<td style="text-align:left;">
Area harvested (Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
101900.00
</td>
<td style="text-align:left;">
Official data
</td>
</tr>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Canary seed
</td>
<td style="text-align:left;">
Yield (Hg/Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
12161.92
</td>
<td style="text-align:left;">
Calculated data
</td>
</tr>
<tr>
<td style="text-align:left;">
Canada
</td>
<td style="text-align:left;">
Grain, mixed
</td>
<td style="text-align:left;">
Area harvested (Ha)
</td>
<td style="text-align:right;">
2012
</td>
<td style="text-align:right;">
57900.00
</td>
<td style="text-align:left;">
Official data
</td>
</tr>
</tbody>
</table>
<p>There are many ways in which we may want to summarize the data table. We could, for example, want to compute the total Barley yield for Canada by year for the years ranging from 2005 and 2007. In R, this would be done in just a few lines of code:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="kw">library</span>(dplyr) </span>
<span id="cb1-2"><a href="#cb1-2"></a>dat2 &lt;-<span class="st"> </span>fao <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="st">    </span><span class="kw">filter</span>(Information <span class="op">==</span><span class="st"> &quot;Yield (Hg/Ha)&quot;</span>,  Crop<span class="op">==</span><span class="st">&quot;Barley&quot;</span>, Country<span class="op">==</span><span class="st">&quot;Canada&quot;</span>,</span>
<span id="cb1-4"><a href="#cb1-4"></a>           Year <span class="op">&gt;=</span><span class="st"> </span><span class="dv">2005</span>, Year <span class="op">&lt;=</span><span class="dv">2010</span>) <span class="op">%&gt;%</span><span class="st">  </span></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="st">    </span><span class="kw">group_by</span>(Year) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="st">    </span><span class="kw">summarise</span>(<span class="dt">Barley_yield =</span> <span class="kw">round</span>(<span class="kw">median</span>(Value))) </span></code></pre></div>
<table>
<thead>
<tr>
<th style="text-align:right;">
Year
</th>
<th style="text-align:right;">
Barley_yield
</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:right;">
2005
</td>
<td style="text-align:right;">
32134
</td>
</tr>
<tr>
<td style="text-align:right;">
2006
</td>
<td style="text-align:right;">
29703
</td>
</tr>
<tr>
<td style="text-align:right;">
2007
</td>
<td style="text-align:right;">
27476
</td>
</tr>
<tr>
<td style="text-align:right;">
2008
</td>
<td style="text-align:right;">
33646
</td>
</tr>
<tr>
<td style="text-align:right;">
2009
</td>
<td style="text-align:right;">
32620
</td>
</tr>
<tr>
<td style="text-align:right;">
2010
</td>
<td style="text-align:right;">
31859
</td>
</tr>
</tbody>
</table>
<p>Creating the same output in a spreadsheet environment would take a bit more effort and its workflow would be less transparent.</p>
</div>
<div id="reproducible-analysis" class="section level2">
<h2>Reproducible analysis</h2>
<p>Data table manipulation is inevitable in any data analysis workflow and, as discussed in the last section, can be prone to clerical errors if performed in a point-and-click environment. Furthermore, reproducing a workflow in a spreadsheet environment can be difficult unless each click and each copy-and-paste are meticulously documented. And even if the documentation is adequate, there is no way of knowing if the analyst followed those exact procedures (unless his mouse and keyboard moves were recorded). However, with a scripting environment, each step of a workflow is clearly and unambiguously laid out. This leads to another basic tenet of the scientific method: <strong>reproducibility of the workflow</strong>.</p>
<p>Reproducible research lends credence to scientific work. The need for reproducibility is not limited to data collection or methodology but includes the actual analytical workflow that generated the results including data table manipulation and implementation of statistical procedures.</p>
<p>Data analysis can be complex. Each data manipulation step that requires human interaction is prone to clerical error. But error can also manifest itself in faulty implementation of an analytical procedure—both technical and theoretical. Unfortunately, the workflows are seldom available in technical reports or peer-reviewed publications where the intended audience is only left with the end product of the analysis.</p>
<div style="width:600px;height:160px;margin-left:70px;margin-bottom:10px;font-family:Garamond, Georgia, serif;font-size:1.5em;font-style:italic">
<p><img style="float:left;margin-right:10px;" src="img/Baggerly_Coombes.png"> “… a recent survey of 18 quantitative papers published in Nature Genetics in the past two years found reproducibility was not achievable even in principle for 10.”</p>
<p>–Keith A. Baggerly &amp; Kevin R. Coombes<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a></p>
</div>
<p>Unfortunately, examples of irreproducible research are <a href="http://theconversation.com/science-is-in-a-reproducibility-crisis-how-do-we-resolve-it-16998">all too common</a>. An example of such was reported by the <a href="http://www.nytimes.com/2011/07/08/health/research/08genes.html?_r=0">New York Times</a> in an article titled <em>How Bright Promise in Cancer Testing Fell Apart</em>. In 2006, researchers at Duke had published a paper in <em>Nature Medicine</em> on a breakthrough approach to fighting cancer. The authors’ research suggested that genomic tests of a cancer cell’s DNA could be used to target the most effective chemotherapy treatment. This was heralded as a major breakthrough in the fight against cancer. Unfortunately, the analysis presented by the authors was flawed. Two statisticians, Dr. Baggerly and Dr. Coombes, sought to replicate the work but discovered instead that the published work was riddled with problems including mis-labeling of genes and confounding experimental designs. The original authors of the research did not make the analytical workflow available to the public thus forcing the statisticians to scavenge for the original data and techniques. It wasn’t until 5 years later, in 2011, that <em>Nature</em> decided to retract the paper because they were “unable to reproduce certain crucial experiments”.</p>
<p>Many journals now require or <em>strongly encourage</em> authors to <em>“make materials, data and associated protocols promptly available to readers without undue qualifications”</em> (<a href="http://www.nature.com/authors/policies/availability.html?message=remove">Nature, 2014</a>). Sharing data file is not too difficult, but sharing the analytical workflow used to generate conclusions can prove to be difficult if the data were run though many different pieces of software and point-and-click procedures. An ideal analytical workflow should be scripted in a human readable way from beginning (the moment the data file(s) is/are read) to the generation of the data tables or data figures used in the report of publication. This has two benefits: elimination of clerical errors (associated with poorly implemented point-and-click procedures) and the exposition of the analytical procedures adopted in the workflow.</p>
</div>
</div>
<div id="creating-dynamic-documents-using-r-markdown" class="section level1">
<h1>Creating dynamic documents using R Markdown</h1>
<p>Another source of error in the write-up of a report or publication is the linking of tables, figures and statistical summaries to the write-up. Typically, one saves statistical plots as image files then loads the image into the document. However, the figures may have gone through many different iterations resulting in many different versions of the image file in a working folder. Add to this many other figures, data table files and statistical results from various pieces of software, one quickly realizes the potential for embedding the wrong image files in the document or embedding the wrong statistical summaries in the text. Furthermore, the researcher is then required to properly archive and document the provenance of each figure, data table or statistical summary resulting in a complex structure of files and directories in the project folder thus increasing the odds of an irreproducible analysis.</p>
<p>Confining all of the analysis to a scripting environment such as R can help, but this still does not alleviate the possibility of loading the wrong figure into the document, or forgetting to update a statistical summary in the text when the original data file was revised. A solution to this potential pitfall is to embed the actual analysis and graphic generation process into the document–such environments are called dynamic documents. In this course, we will use the <a href="http://rmarkdown.rstudio.com/">R Markdown authoring tool</a> which embeds R code into the document. An example of an R Markdown document is this write-up where both the CO2 data analysis and the data manipulation examples were generated using R and whose code was embedded in this R Markdown document. You can view the contents of the R Markdown file <a href="Week01.Rmd">here</a>.</p>
</div>
<div class="footnotes">
<hr />
<ol>
<li id="fn1"><p>Tukey, John W. <em>Exploratory Data Analysis</em>. 1977. Addison-Wesley.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Cleveland, William S. <em>Visualizing Data</em>. 1993. Hobart Press.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>Lewin-Koh, Nicholas and Theus, Martin (Eds). <em>Statistical Graphics and InfoVis: separated Twins at Birth?</em> ASA Statistical Computing and Graphics Newsletter, 2011.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>Baggerly, Keith A. and Coombes, Kevin R. <em>Deriving Chemosensitivity from Cell Lines: Forensic Bioinformatics and Reproducible Research in High-Throughput Biology</em>. The Annals of Applied Statistics, vol.3, no.4, pp. 1309-1334. 2009.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>


<div class="footer">
<hr/>
<a rel="license" href="https://creativecommons.org/licenses/by-nc/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc/4.0/80x15.png" /></a>  Manny Gimond (2020)
</br>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = false;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
